Firestore/core/src/firebase/firestore/core/relation_filter.cc

```diff
@@ -41,48 +67,76 @@ const FieldPath& RelationFilter::field() const {
 
 bool RelationFilter::Matches(const model::Document& doc) const {
   if (field_.IsKeyFieldPath()) {
-    // TODO(rsgowman): Port this case
-    abort();
+    HARD_ASSERT(value_rhs_.type() == FieldValue::Type::Reference,
+                "Comparing on key, but filter value not a Reference.");
+    HARD_ASSERT(op_ != Filter::Operator::ArrayContains,
+                "arrayContains queries don't make sense on document keys.");
+    const auto& ref = value_rhs_.reference_value();
+    ComparisonResult comparison = doc.key().CompareTo(ref.key());
+    return MatchesComparison(comparison);
   } else {
     absl::optional<FieldValue> doc_field_value = doc.field(field_);
     return doc_field_value && MatchesValue(doc_field_value.value());
   }
 }
 
-bool RelationFilter::MatchesValue(const FieldValue& other) const {
+bool RelationFilter::MatchesValue(const FieldValue& lhs) const {
   if (op_ == Filter::Operator::ArrayContains) {
-    if (other.type() != FieldValue::Type::Array) return false;
+    if (lhs.type() != FieldValue::Type::Array) return false;
 
-    const std::vector<FieldValue>& contents = other.array_value();
+    const auto& contents = lhs.array_value();
     return absl::c_linear_search(contents, value_rhs_);
```

Comment:
Question: do you know the idea behind using `c_linear_search`? From the implementation, it simply delegates to `std::find`.

Reply:
Ah, I see, it's basically `contains`, except with a non-intuitive name.

Reply:
It's an un-intuitive name that goes out of its way to imply that it's O(n).